File: leetcode/347-TopKFrequentElements/347-TopKFrequentElements.cpp
// Last updated: 29/04/2026, 23:10:38
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        vector<int> result;
        unordered_map<int, int> counter;

        for (int i = 0; i < nums.size(); i++) {
            counter[nums[i]]++;
        }

        map<int, vector<int>> occurenceCounter;
        for (auto x: counter) {
            occurenceCounter[x.second].push_back(x.first);
        }

        k = counter.size() - k;
        for (auto x: occurenceCounter) {
            for (int i = 0; i < x.second.size() && k >= 0; i++) {
                if (k == 0) {
                    result.push_back(x.second[i]);
                } else {
                    k--;
                }
            }
        }

        return result;
    }
};

File: leetcode/347-TopKFrequentElements/347-TopKFrequentElements.java
// Last updated: 29/04/2026, 23:10:32
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> m = new HashMap<>();
        for(int num: nums) {
            m.putIfAbsent(num, 0);
            m.put(num, m.get(num)+1);
        }
        PriorityQueue<Map.Entry<Integer, Integer>> pq = new PriorityQueue<>((a,b)->b.getValue()-a.getValue());
        pq.addAll(m.entrySet());
        int[] response = new int[k];
        for(int i=0;i<response.length; i++) {
            response[i] = pq.poll().getKey();
        }
        return response;
    }
}